lib/semantics/check-do-concurrent.cc

```diff
@@ -0,0 +1,270 @@
+// Copyright (c) 2018, NVIDIA CORPORATION.  All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "check-do-concurrent.h"
+#include "attr.h"
+#include "scope.h"
+#include "symbol.h"
+#include "../parser/message.h"
+#include "../parser/parse-tree-visitor.h"
+
+namespace Fortran::semantics {
+
+using namespace parser::literals;
+
+// 11.1.7.5 - enforce semantics constraints on a DO CONCURRENT loop body
+class DoConcurrentEnforcement {
+public:
+  DoConcurrentEnforcement(parser::Messages &messages) : messages_{messages} {}
+  std::set<parser::Label> labels() { return labels_; }
+  template<typename T> bool Pre(const T &) { return true; }
+  template<typename T> void Post(const T &) {}
+  template<typename T> bool Pre(const parser::Statement<T> &statement) {
+    charBlock_ = statement.source;
+    if (statement.label.has_value()) {
+      labels_.insert(*statement.label);
+    }
+    return true;
+  }
+  // C1136
+  void Post(const parser::ReturnStmt &) {
+    messages_.Say(charBlock_,
+        parser::MessageFormattedText{
```

Comment:
Is the construction of a `parser::MessageFormattedText` necessary here? A `Message` can be constructed from a `MessageFixedText` so I think you can use just `messages_.Say(charBlock_, "..."_err_en_US);`.

Reply:
ok